miscellaneous patches supporting thread protection

tyrantlizard77 opened the pull request:
- privatize some initialization macros that were not intended to be public
- init timer driver earlier in boot sequence
- fix a semaphore accessed before first use
- fix alignment bug with kernel sections used as arrays of kernel objects
- Increase QEMU target default RAM to 8MB
- add MMU_LIST debug info to gen_mmu.py
- Don't do no-execute tests on i386 (no HW support)

Commit:
kernel: make K_.*_INITIALIZER private to kernel

Upcoming memory protection features will be placing some additional
constraints on kernel objects:

- They need to reside in memory owned by the kernel and not the
application
- Certain kernel object validation schemes will require some run-time
initialization of all kernel objects before they can be used.

Per Ben these initializer macros were never intended to be public. It is
not forbidden to use them, but doing so requires care: the memory being
initialized must reside in kernel space, and extra runtime
initialization steps may need to be peformed before they are fully
usable as kernel objects. In particular, kernel subsystems or drivers
whose objects are already in kernel memory may still need to use these
macros if they define kernel objects as members of a larger data
structure.

It is intended that application developers instead use the
K_<object>_DEFINE macros, which will automatically put the object in the
right memory and add them to a section which can be iterated over at
boot to complete initiailization.

There was no K_WORK_DEFINE() macro for creating struct k_work objects,
this is now added.

k_poll_event and k_poll_signal are intended to be instatiated from
application memory and have not been changed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
linker/common-ram.ld: force input section align

These special kernel sections represent arrays of kernel objects than
are iterated over at runtime to perform initialization.

The code expects all the data in these sections to be in the form of an
array of that section type, with each element sizeof(type) bytes apart.

Unfortunately, the linker sometimes has other plans and in some cases
was defaulting to aligning the data to some large power-of-two value,
such as 64 bytes. This causes any attempt to iterate over these sections
to fail as they are not a proper array.

Use the ld SUBALIGN() directive to force the alignment of these input
sections to 4 bytes.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: timer: init earlier in boot sequence

By the time we get to POST_KERNEL, kernel services and kernel objects
should be available for use. This should include timers and the random
number generator, but we don't init the system clock until sometime
during the POST_KERNEL phase. Initialize it earlier.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: net: route: fix semaphore usage

A network stack thread was trying to use a semaphore before it was
initialized. Set it up at toplevel so it is ready to go.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

dbkinder reviewed (dismissed):
+1 for doc change  :)

jukkar commented on samples/net/common/sample_app_setup.c:
```diff
@@ -31,7 +31,7 @@
 #endif
 
-static struct k_sem waiter = K_SEM_INITIALIZER(waiter, 0, 1);
+static K_SEM_DEFINE(waiter, 0, 1);
```
This file was removed few days ago, and its functionality was moved into subsys/net/lib/app/init.c

Commit:
qemu_x86: increase to 8MB of RAM

Previously we were instantiating QEMU with 32MB of RAM but
only enabling a small fraction of it.

Now we boot with 8MB of ram. We ignore the first 4K so we can
make that an unmapped paged to catch NULL pointer dereferences.

If XIP is enabled, the "ROM" region will be the first half of
memory, the "RAM" region the latter.

Move the IDT_LIST and MMU_LIST regions elsewhere so they don't
overlap the new memory arrangement.

Use !XIP to fix a problem where CONFIG_RAM_SIZE was set incorrectly
for XIP case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: net: remove overriden RAM size

The QEMU targets have plenty of RAM now and the current value
specified wasn't a multiple of a 4K page.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: protection: don't do exec tests on x86

The IA32 MMU has no concept of a "no execute" flag, this is
unfortunately only implemented in x86_64.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
mmustructs: don't use addr to name structures

This breaks too easily, for example if &some_linker_variable
is used. The names don't matter at all, use preprocessor
__COUNTER__.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
gen_mmu.py: verbose mode dump MMU_LIST

It's useful to see what original memory regions were configured
in code via the MMU_BOOT_REGION() macros.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

jhedberg reviewed (commented):
Not being able to place kernel objects as members of parent objects prevents object-oriented design and will likely discourage their usage in apps. In the absence of this ability apps will likely start re-inventing their own objects. I think that's a very high price to pay, so I hope the benefits have been carefully thought out and are worth it.

andrewboie commented:
I don't feel there is any way around this.
Kernel objects contain data that is private to the kernel, if a non-privileged thread corrupts or maliciously modifies this data, the whole system can go down, not just the offending thread.
Drivers, subsystems, and privileged application code can continue to have hierarchical structures of kernel objects so long as they are placed and initialized correctly.
But unprivileged application code, the kernel objects absolutely must be opaque.

andrewboie commented:
@jhedberg also worth mentioning: this really only impacts code that was specifically using the K_*_INITIALIZER macros. Code that puts kernel objects in a struct, and then makes a runtime call to initialize it should work as before. The key thing here is that in some memory protection schemes, it is not possible to completely initialize a kernel object at build time, so we don't want K_*_INITIALIZER used unless the user knows exactly what they are doing.

jhedberg commented:
Ok, if the run-time initialization is still possible then that takes care of my concerns. However, then it's no longer accurate to say that the objects must always reside in kernel owned memory.

andrewboie commented:
> Ok, if the run-time initialization is still possible then that takes care of my concerns. However, then it's no longer accurate to say that the objects must always reside in kernel owned memory.

I'm not following you here.
System calls will reject any kernel object pointers that don't fall within the range of memory for the kernel, if CONFIG_APPLICATION_MEMORY is enabled and an application tries to use a kernel object allocated in its memory range. Applications defining kernel objects will either need to use the K_<object>_DEFINE macros, use a __kernel attribute when declaring them, or use runtime allocation APIs for creating kernel objects.

Applications never write to kernel objects directly, this is all handled via system calls which do necessary privilege elevation.

If CONFIG_APPLICATION_MEMORY is not enabled, then all memory is kernel memory, and unprivileged threads will need to use APIs to specifically request access to memory outside the thread's stack. This is what FreeRTOS-MPU does.

andyross commented on arch/x86/include/mmustructs.h:
```diff
@@ -129,17 +129,26 @@ struct mmu_region {
-		.address = addr,					\
-		.size = region_size,					\
+		.address = (u32_t)(addr),				\
+		.size = (u32_t)(region_size),				\
```
Typesafety whine: Seems like if the user is passing an incompatible type to the macro, that's an error we want flagged and not suppressed.   the region_size argument should 100% absolutely be a size_t value already or it's a bug.  It seems like maybe you want to define the address field as a void* and not an integer so you can typecheck the macro, but maybe that's an API we don't want to modify.

andrewboie replied:
a lot of the time we are gonna get the size from taking the address of a linker variable. but its not a big deal to make that an explicit cast, I'll remove these.

andyross commented on tests/kernel/protection/src/main.c:
```diff
@@ -48,11 +48,17 @@ void _SysFatalErrorHandler(unsigned int reason, const NANO_ESF *pEsf)
+/* i386 MMU doesn't control execute capabilities, only on x86_64 */
+#ifndef CONFIG_X86
+#define NO_EXECUTE_SUPPORT 1
+#endif
```
Whine: seems like the proper magic would be to enumerate "NX_SUPPORT" or whatever as a predicate macro instead of assuming that every architecture other than i386 supports it.

andrewboie replied:
sure I can make this an affirmative rather than negative check.

andyross reviewed (approved):
Nothing but minor nitpicking

lpereira reviewed (dismissed):
LGTM.

andrewboie commented:
recheck

andrewboie commented:
I just did a rebase to try to get Shippable un-stuck. Doesn't seem to be working....

Commit:
qemu: enable MMU by default

We have lots of RAM, this helps catch bugs.
Enable XIP as well, this used to be turned on but was
shut off for some reason.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
